src/Tools/Source/RunTests/TestRunner.cs

```diff
@@ -40,6 +42,95 @@ internal TestRunner(Options options, ProcessTestExecutor testExecutor)
+        {
+            var sourceBranch = Environment.GetEnvironmentVariable("BUILD_SOURCEBRANCH");
+            if (sourceBranch is null)
+            {
+                ConsoleUtil.WriteLine("BUILD_SOURCEBRANCH is not set");
+                sourceBranch = "local";
+                Environment.SetEnvironmentVariable("BUILD_SOURCEBRANCH", sourceBranch);
+            }
+
+            var msbuildTestPayloadRoot = ".";
+            var isAzureDevOpsRun = Environment.GetEnvironmentVariable("SYSTEM_ACCESSTOKEN") is not null;
+            if (!isAzureDevOpsRun)
+            {
+                ConsoleUtil.WriteLine("SYSTEM_ACCESSTOKEN environment variable was not set, so test results will not be published.");
+                msbuildTestPayloadRoot = "$(RepoRoot)artifacts/testPayload";
+            }
+            var correlationPayload = $@"<HelixCorrelationPayload Include=""{msbuildTestPayloadRoot}/.duplicate"" />";
+
+            if (Environment.GetEnvironmentVariable("BUILD_REPOSITORY_NAME") is null)
+                Environment.SetEnvironmentVariable("BUILD_REPOSITORY_NAME", "dotnet/roslyn");
+
+            if (Environment.GetEnvironmentVariable("SYSTEM_TEAMPROJECT") is null)
+                Environment.SetEnvironmentVariable("SYSTEM_TEAMPROJECT", "dnceng");
+
+            if (Environment.GetEnvironmentVariable("BUILD_REASON") is null)
+                Environment.SetEnvironmentVariable("BUILD_REASON", "pr");
+
+            var buildNumber = Environment.GetEnvironmentVariable("BUILD_BUILDNUMBER") ?? "0";
+            var workItems = assemblyInfoList.Select(ai => makeHelixWorkItemProject(ai));
+            var project = @"
+<Project Sdk=""Microsoft.DotNet.Helix.Sdk"" DefaultTargets=""Test"">
+    <PropertyGroup>
+        <HelixSource>pr/" + sourceBranch + @"</HelixSource>
+        <HelixType>test</HelixType>
+        <HelixBuild>" + buildNumber + @"</HelixBuild>
+        <HelixTargetQueues>" + _options.HelixQueueName + @"</HelixTargetQueues>
+        <Creator>rigibson</Creator>
+        <IncludeDotNetCli>true</IncludeDotNetCli>
+        <DotNetCliPackageType>sdk</DotNetCliPackageType>
+        <EnableAzurePipelinesReporter>" + (isAzureDevOpsRun ? "true" : "false") + @"</EnableAzurePipelinesReporter>
+    </PropertyGroup>
+
+    <ItemGroup>
+        " + correlationPayload + string.Join("", workItems) + @"
+    </ItemGroup>
+</Project>
+";
+
+            File.WriteAllText("helix-tmp.csproj", project);
+            var process = ProcessRunner.CreateProcess(
+                executable: _options.DotnetFilePath,
+                arguments: "build helix-tmp.csproj",
+                captureOutput: true,
+                onOutputDataReceived: (e) => ConsoleUtil.WriteLine(e.Data),
+                cancellationToken: cancellationToken);
+            var result = await process.Result;
+
+            // TODO: how do we handle publishing stuff like proc dumps when test runs have crashes?
```

Comment:
Have we asked core-eng how this should be handled?

Reply:
It sounds like Helix machines automatically set up the configuration for generating dumps when the test process crashes. I'll just play with adding crashes to the tests and see if I can get them to publish and see if we need to do anything else.

Reply:
Here's an example PR which publishes dumps: https://dev.azure.com/dnceng/public/_build/results?buildId=949005&view=ms.vss-test-web.build-test-results-tab&runId=29928386&paneView=debug&resultId=210588

Note that in subsequent commits I did actually add the name of the job to the helix run name, so that the test failures list is more comprehensible.

Reply:
Nice. 

Still though I think we need to track down the mechanism by which this is occurring. There are a couple of different ways you can go about this and it would impact the situations in which we do or do not get dumps. Want to make sure they're doing this in a way that will get all the dumps we are accustomed to.